baseplate/server/queue_consumer.py

```diff
@@ -0,0 +1,268 @@
+        ]
+        return cls(
+            pump=consumer_factory.build_pump_worker(work_queue),
+            handlers=handlers,
+            healthcheck_server=consumer_factory.build_health_checker(listener),
+        )
+
+    def _terminate(self) -> None:
+        """Send a SIGTERM signal to ourselves so baseplate can call `stop`."""
+        assert self.started
+        if not self.stopped:
+            os.kill(os.getpid(), signal.SIGTERM)
+
+    def _watch(self) -> None:
+        """Watches the threads to see if they have stopped prematurely.
+
+        Loops through the pump thread and handler threads periodically to check
+        if they are still alive and terminates the server if one is no longer
+        alive.
+        """
+        assert self.started
+        watched = [self.pump_thread] + self.threads
+        while not self.stopped:
+            for thread in watched:
+                if not thread.is_alive():
+                    logger.debug("Dead thread found, terminating server.")
+                    self._terminate()
+                    return
+            time.sleep(self._watch_period)
+
+    def start(self) -> None:
+        """Start the server.
+
+        Starts all of the worker threads and the healthcheck server (if it exists).
+
+        Should only be called once and should not be called after the server is
+        stopped, will raise an AssertionError in either of those cases.
+        """
+        assert not self.started
+        assert not self.stopped
+        logger.debug("Starting server.")
+        self.started = True
+        logger.debug("Starting pump thread.")
+        self.pump_thread.start()
+        logger.debug("Starting message handler threads.")
+        for thread in self.threads:
+            thread.start()
+        logger.debug("Starting thread terminator.")
+        Thread(target=self._watch, daemon=True).start()
+        logger.debug("Starting healthcheck server.")
+        self.healthcheck_server.start()
+        logger.debug("Server started.")
+
+    def stop(self) -> None:
+        """Start the server.
+
+        Stop all of the worker threads and the healthcheck server (if it exists).
+        Waits for the handler threads to drain before returning but does not wait
+        for the pump or watcher threads to finish.
+
+        Should only be called once and should not be before the server is
+        started, will raise an AssertionError in either of those cases.
+        """
+        assert self.started
+        assert not self.stopped
+        logger.debug("Stopping server.")
+        self.stopped = True
+        logger.debug("Stopping pump thread.")
+        self.pump.stop()
+        logger.debug("Stopping message handler threads.")
+        for handler in self.handlers:
+            handler.stop()
+        logger.debug("Waiting for message handler threads to drain.")
+        for thread in self.threads:
+            thread.join()
+        logger.debug("Stopping healthcheck server.")
```

Comment:
is there any ordering requirement here? is there any for the startup?

Reply:
As this is currently written, yes for startup (you want to start the watcher thread after all of the other threads it is meant to watch).

For shutdown, the ordering of `handler.stop` before `thread.join` is important (otherwise you'll be waiting forever on threads that haven't been stopped) but the rest is _less_ important. Stopping the pump first is nice so you stop consuming messages from your external message queue and stopping the healthchecker last is probably a good idea in case you have a system that aggressively kills processes that fail their health checks.